docs/design/018-new-http-api.md

```diff
@@ -0,0 +1,486 @@
+}
+```
+
+- Passing a socket with custom transport settings, some HTTP options, and making a POST request:
+```javascript
+import { TCP } from 'k6/x/net';
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const socket = await TCP.open('10.0.0.10:80', { keepAlive: true });
+  const client = new Client({
+    socket: socket,
+    proxy: 'https://myproxy',
+    version: 1.1,                     // force a specific HTTP version
+    headers: { 'User-Agent': 'k6' },  // set some global headers
+  });
+  await client.post('http://10.0.0.10/post', {
+    json: { name: 'k6' }, // automatically adds 'Content-Type: application/json' header
+  });
+}
+```
+
+- A tentative HTTP/3 example:
+```javascript
+import { UDP } from 'k6/x/net';
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const socket = new UDP.open('192.168.1.1:9090');
+
+  const client = new Client({
+    socket: socket,
+    version: 3,      // A UDP socket would imply HTTP/3, but this makes it explicit.
+  });
+  await client.get('https://httpbin.test.k6.io/get');
+}
+```
+
+
+#### Host name resolution
+
+Host names can be resolved to IP addresses in several ways:
+
+- Via a static lookup map defined in the script.
+- Via the operating system's facilities (`/etc/hosts`, `/etc/resolv.conf`, etc.).
+- By querying specific DNS servers.
+
+When connecting to an address using a host name, the resolution can be controlled via the `lookup` function passed to the socket constructor. By default, the mechanism provided by the operating system is used (`dns.lookup()`).
+
+For example:
+```javascript
+import { TCP } from 'k6/x/net';
+import dns from 'k6/x/net/dns';
+
+const hosts = {
+  'hostA': '10.0.0.10',
+  'hostB': '10.0.0.11',
+};
+
+export default async function () {
+  const socket = await TCP.open('myhost', {
+    lookup: async hostname => {
+      // Return either the IP from the static map, or do an OS lookup,
+      // or fallback to making a DNS query to specific servers.
+      return hosts[hostname] || await dns.lookup(hostname) ||
+        await dns.resolve(hostname, {
+          rrtype: 'A',
+          servers: ['1.1.1.1:53', '8.8.8.8:53'],
+        });
+    },
+  });
+}
+```
+
+#### Requests and responses
+
+HTTP requests can be created declaratively, and sent only when needed. This allows reusing request data to send many similar requests.
+
+For example:
+```javascript
+import { Client, Request } from 'k6/x/net/http';
+
+export default async function () {
+  const client = new Client({
+    headers: { 'User-Agent': 'k6' },  // set some global headers
+  });
+  const request = new Request('https://httpbin.test.k6.io/get', {
+    // These will be merged with the Client options.
+    headers: { 'Case-Sensitive-Header': 'somevalue' },
+  });
+  const response = await client.get(request, {
+    // These will override any options for this specific submission.
+    headers: { 'Case-Sensitive-Header': 'anothervalue' },
+  });
```

Comment:
This API proposal seems a bit messy. The HTTP request method should probably be part of the `Request` object, right? The whole point of a `Request` object is to fully contain everything you need to make a request.

But then, what `client.get()` probably shouldn't accept `Request` parameters, since that `Request` could be a `POST`. I am not sure if methods like `.get()` and `.post()` are even necessary in the default HTTP `Client` API :thinking: Maybe a single generic `Client.request()` or `Client.do()` is enough. But if we choose to have helper methods, they probably shouldn't accept a `Request`, just a `method, params, body` or something like that :thinking: